src/gui/WinMain.cpp

```diff
@@ -77,7 +82,10 @@ int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine
 
 	if (bHasLoadArgument) {
 #ifndef CXBX_LOADER
-		CxbxKrnlMain(__argc, __argv);
+		uint32_t SystemDevBlocksReserved[384] = {};
+		int system = SYSTEM_XBOX; // TODO: Temporary placeholder until loader is functional.
+		ReserveAddressRanges(system, SystemDevBlocksReserved);
+		CxbxKrnlMain(__argc, __argv, SystemDevBlocksReserved);

```

Comment:
This is fine, however the changes I have made to the VMManager class make it incompatible with the old approach, so, non-loader builds are essentially broken now.

Reply:
I'd be good if both approaches could be kept in working order, since that would allow earlier delivery of this branch. (We've had bad experiences before with branches that stayed unmerged for too long)

Reply:
We can still merge this to the loader branch in the main repository, so that develop will not be affected. Also, there are a lot of places where the VMManager assumes a loader has reserved the memory. To keep both approaches functional would require to reinstate all the old code plus lots of `ifdef`, which would be a huge waste in my opinion since the old code is going to be deleted anyway when this branch is functional.

Reply:
I'll take your word for it without hesitation. It would have been nice, but keeping this in a separate branch on the main repository will have to suffice then. Cheers

Reply:
I'd had hoped it was possible to update the memory manager code in such a way that both approached stayed in working order, but since you're doing all the hard work, I can hardly complain, now can I? 😜

Reply:
> Also, there are a lot of places where the VMManager assumes a loader has reserved the memory.

Oh? Isn't it `ReserveAddressRanges` suppose to do reserved the memory? Basically, should be identical to loader's method except includes 128MB pre-allocated memory within cxbx.exe. Unless I'm missing some calls.

Reply:
In the old apprroch the devkit and system regions were not reserved so the VMManager could allocate there directly. With the loader, those regions are reserved so all the functions that allocate there should only commit memory and also in 64KiB blocks (this restriction wasn't present before). If, as I understand it, you plan to use the loader approach also with the non-loader path to keep that happy and merge in develop branch, it will be broken since the loader still doesn't work now.

Reply:
I'm sorry, there shouldn't be any misunderstanding here; My initial idea for this branch was (and ideally, still is): With the same code, build both the current approach (having GUI+loader+emulator in the same executable), and the new  approach (wherein all three are separate executables)

Reply:
This, using ifdefs in source code that is compiled in both approaches but must behave differently, so that we can run either the current all-in-one cxbx.exe (doing everything it's been doing up until now), or use the new standalone GUI, that launches the loader (which reserves memory and loads the emulation DLL for actual emulation).

Reply:
Of course, once the new approach is working as well or better then the old all-in-one approach, we can make the switch and remove all then-obsolete code

Reply:
IMHO, once loader project is working. Then we should make the launch switch to the loader executable from existing GUI, not from the new GUI template which cba to redo everything when existing GUI is still functional.

Reply:
True, we could optionally launch the loader from the old GUI. But unless I'm mistaken, I already constructed the new GUI project using the same user-interface source code files as the existing all-in-one cxbx.exe project. Originally it was my plan to have this stripped GUI launch the loader, while the old GUI kept on launching itself with a `/load path.xbe` command line argument.

Reply:
For this to work, the existing (shared between old and new) GUI code that launches emulation should also differentiate between old and new approach using an ifdef.

Reply:
The main goal is actually remove the 64MB variable which means the emulation within the GUI can be no longer used. Though, I don't see the real reason to must have 2 separate GUIs nor launch the emulation within the GUI. `cxbxr-emu.dll` is the real work for the emulation which later on become a class like @LukeUsher originally planned for to able use all-in-one executable + third-party front-end usage. Although, it is possible to use `cxbxr-emu.dll` to handle everything including create/delete the xbox emulation class.

Though, we can still add an option to toggle launch to GUI emulation or loader's emulation within existing GUI for **temporary purpose**.

Plus `cxbxr.exe` is originally intended to use with Qt's GUI.

Reply:
Yeah, the 'new' GUI project is basically the same GUI, just stripped from emulation code. It can certainly be replaced by a better one in the future. The reason I created a second project for it, was to show what the end result would roughly look like.

Reply:
But the goal was not only to remove the memory placeholder, but primarily to have maximal control over all required memory ranges

Reply:
The new GUI project wouldn't help at all. Since we're still using the emulation functions from GUI side unfortunately. The best thing we can do is relocate those functions into common folder. Plus there are GUI functions also in placed in the emulation folder. For one example is the game logo to display on GUI's bottom right corner, like @LukeUsher mentioned shouldn't be in emulation's folder.

Reply:
Once the above are done, we can remove the "emu" list of files macro from the existing GUI's cmake file.

Reply:
> But the goal was not only to remove the memory placeholder, but primarily to have maximal control over all required memory ranges

Which should be the loader's responsibility atm. 😉 The existing GUI isn't up to the task able do that. So... 🤷‍♂

Once we do have virtual emulation support (eventually), then we don't need loader project anymore.

Reply:
It seems we all agree now; the old GUI should eventually go, the loader-approach should be brought into a working state, code should reside in the correct folders, unwanted dependencies should be removed, etc.

It's just a lot of work to get there...

Reply:
Let's continue the discussion on discord if there are any. 😉